R/utils-get_code_dependency.R

```diff
@@ -0,0 +1,267 @@
+
+      # Watch out for SYMBOLS after $ and @. For x$a x@a: x is object, a is not.
+      # For x$a, a's ID is $'s ID-2 so we need to remove all IDs that have ID = $ID - 2.
+      dollar_ids <- x[x$token %in% c("'$'", "'@'"), "id"]
+      if (length(dollar_ids)) {
+        object_ids <- x[sym_cond, "id"]
+        after_dollar <- object_ids[(object_ids - 2) %in% dollar_ids]
+        sym_cond <- setdiff(sym_cond, which(x$id %in% after_dollar))
+      }
+
+      # If there was an assignment operation detect direction of it.
+      ass_cond <- grep("ASSIGN", x$token)
+      if (length(ass_cond)) { # NOTE(1)
+        sym_cond <- sym_cond[sym_cond > ass_cond] # NOTE(2)
+      }
+      if ((length(ass_cond) && x$text[ass_cond] == "->") || !length(ass_cond)) { # NOTE(3)
+          sym_cond <- rev(sym_cond)
+      }
+      append(x[sym_cond, "text"], ":", after = 1)
+
+      ### NOTE(3): What if there are 2+ assignments, e.g. a <- b -> c or e.g. a <- b <- c.
+      ### NOTE(2): For cases like 'eval(expression(b <- b + 2))' removes 'eval(expression('.
+      ### NOTE(1): Cases like 'data(iris)' that do not have an assignment operator.
+      ### NOTE(1): Then they are parsed as c("iris", ":", "data")
+    }
+  )
+}
+
+#' Extract Side Effects
+#'
+#' @description Extract all object names from the code that are marked with `@linksto` tag.
+#'
+#' @details The code sometimes consists of functions calls that create side effects. Those are actions that modify the
+#' environment outside of the function environment. Sometimes there is no way to understand, just based on static code
+#' analysis, which objects are modified or created by a function call that has side effects. To allow to point to
+#' objects that are modified or created by function calls with side effects, a `@linksto` comment tag was introduced.
+#' This tag enables to restore the reproducible code for an object, with all side effects that do not directly change
+#' this object. Read more about side effects and the usage of `@linksto` tag in [`get_code_dependencies()`] function.
+#'
+#' @param calls_pd A `list` of `data.frame`s, which is a result of `utils::getParseData()` grouped into separate calls.
+#' A result of `extract_calls()` function.
+#' @return A `list` of length equal to the lenght of `calls_pd` of character vectors of names of objects that are
+#' influenced by `@linksto` tag in a corresponding `call` element of `calls_pd`.
+#' @keywords internal
+#' @noRd
+extract_side_effects <- function(calls_pd) {
+  lapply(
+    calls_pd,
+    function(x) {
+      linksto <- grep("@linksto", x[x$token == "COMMENT", "text"], value = TRUE)
+      unlist(strsplit(sub("\\s*#\\s*@linksto\\s+", "", linksto), "\\s+"))
+    }
+  )
+}
+
+#' @keywords internal
+#' @noRd
+remove_graph_duplicates <- function(graph) {
+  # This function is just a nice to have.
+  # This is not crucial for the implementation.
+  # Purpose: remove duplicates before ":" and after, in each element of list.
```

Comment:
This happens on concatenation of `side_effects` and `cooccurrence` in 

`graph <- mapply(c, side_effects, cooccurence, SIMPLIFY = FALSE)`

and duplicates can appear because there could be a line like `a <- fun1(b) # @linksto a` , which means `a` was in `side_effects` and `cooccurrence`

I think we could reduce duplicates in the `cooccurence` part for cases like `a <- fun1(b, b, b)` with this 6347792e7394e4bf8dce5d1b9a58def97e072e07

Reply:
OR play with the `FUN` in `mapply` a little?

Reply:
yeah, could be way simpler than dealing with duplicates!